Refactor inbox to support deduplicated inbox-recipient storage

pfefferle opened the pull request:
This PR refactors the inbox system to support shared inbox functionality where activities are stored once with multiple recipients as metadata entries, rather than creating duplicate posts per recipient. Key changes include updating meta registration to allow multiple `_activitypub_user_id` values, introducing new shared inbox hooks, and adding recipient management methods.

## Proposed changes:

- Introduces `activitypub_inbox_shared` hooks for efficient multi-recipient processing
- Updates inbox collection to store recipients as separate meta entries
- Adds migration to clean up existing inbox items for fresh shared inbox structure

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [X] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Refactored the inbox system to use a shared inbox, storing activities once with multiple recipients for improved efficiency and reduced duplication.

</details>

pfefferle committed:
Refactor inbox to support deduplicated recipient storage

Migrates inbox activities to a deduplicated storage model where each activity can have multiple recipients stored as separate meta entries. Updates post meta registration to allow multiple _activitypub_user_id values, refactors inbox collection logic to handle multiple recipients, and introduces new shared inbox hooks for efficient processing. Legacy per-user hooks are retained for backward compatibility, and migration logic is added to convert existing data. Improves REST controllers to use the new shared recipient model and ensures blocked users are filtered out before processing.

pfefferle committed:
Merge branch 'trunk' into improve-shared-inbox

pfefferle committed:
Refactor inbox handlers to support context parameter

Unified the inbox request handlers to accept a context parameter distinguishing between user and shared inboxes. Updated REST controllers and action hooks to pass context, removed legacy double-processing logic, and improved extensibility for future inbox handling.

pfefferle committed:
Refactor recipient addition in Inbox class

Replaces direct post meta manipulation with the self::add_recipient method for adding recipients to activities, improving code consistency and maintainability.

pfefferle committed:
Refactor inbox action hooks for shared inbox

Moved deprecated per-user action hooks from main request handler to a dedicated trigger_inbox_actions method, now called via the 'activitypub_inbox_shared' action. Updated parameter naming in Inbox handler for clarity and consistency.

pfefferle committed:
Refactor trigger_inbox_actions parameter types and context

Updated trigger_inbox_actions to require array types for $data and $allowed_recipients. Also fixed the context parameter passed to do_action to use the provided $context instead of a hardcoded value.

Jiwoon-Kim commented:
Since this idea is more of a tentative hypothesis, I preferred to discuss it in the comments rather than formally submitting it as an issue.

Summary of the idea:

- Treat activities as objects and store them in the wp_post table.
- Implement Inbox and Outbox using taxonomies.
- Question: Should remote activities and local activities be separated into different CPTs (custom post types)?
- Reason: Remote activities create remote objects, whereas local activities are generated from local objects.

pfefferle committed:
Merge branch 'trunk' into improve-shared-inbox

pfefferle commented:
Inbox/Outbox are already CPTs: https://github.com/Automattic/wordpress-activitypub/blob/trunk/includes/class-post-types.php

Jiwoon-Kim commented:
Yes, since the Inbox and Outbox are implemented as CPTs, I suspected that for local-to-local activities there could be a conflict between the URIs of the activity in the sender's Outbox and the receiver's Inbox.

pfefferle commented:
Shouldn't be an issue!

pfefferle committed:
Merge branch 'trunk' into improve-shared-inbox

Jiwoon-Kim commented:
Yes, I’m glad to hear it’s not an issue. Intuitively, I imagined that a single activity would exist as a duplicate object in both the Inbox and the Outbox. I don’t fully understand why that would be incorrect or why it’s not a problem, but as long as it works properly in practice, that’s what matters.

---

Exactly, that makes sense. Intuitively, it does feel odd to imagine the same activity “existing twice” — once in the sender’s Outbox and once in the receiver’s Inbox — as if it were a duplicated object.

The reason it’s not actually a problem is that in ActivityPub, collections like Outbox and Inbox are just containers or references to activities. The activity itself is identified by its URI, and each actor’s collection merely points to it. So even if the same activity appears in multiple collections, it’s not considered a duplication in a harmful sense — it’s just the protocol keeping track of where the activity is relevant.

As long as the flow works correctly (sending, receiving, processing activities), the system behaves as expected, and you don’t need to worry about “conflict” between Inbox and Outbox entries.

pfefferle committed:
Refactor inbox migration to cleanup all inbox items

Simplifies the inbox migration by removing deduplication and recipient metadata logic. Now deletes all existing inbox items to prepare for a new shared inbox structure.

pfefferle committed:
Rename inbox migration method for clarity

Renamed the method migrate_to_inbox_deduplication to cleanup_inbox and updated its usage to better reflect its purpose of deleting existing inbox items for the new shared inbox structure.

pfefferle committed:
Add middleware to filter shared inbox requests

Introduces maybe_handle_inbox_request to prevent duplicate processing of shared inbox requests by filtering based on context. The inbox handler now uses this middleware, while shared inbox requests are processed directly.

pfefferle committed:
Refine recipient handling and update inbox tests

Recipient filtering now preserves 0 (blog user) and excludes only null/false values. Negative user IDs are rejected, but 0 is allowed in add_recipient. PHPUnit tests updated to use actions instead of filters for inbox handling and to enable inbox persistence during tests.

pfefferle committed:
Update callback signature for activitypub_handled_inbox

Removed the unused $context parameter from the callback function and adjusted the add_action argument count from 6 to 5 to match the new signature.

pfefferle committed:
Allow user_id 0 in remove_recipient method

Updated remove_recipient to permit user_id 0, which represents the blog user, while continuing to reject negative values. This change improves handling of special user cases.

pfefferle committed:
Add shared inbox support and enhance recipient handling

Refactors inbox logic to support multiple recipients and shared inbox context, including normalization of recipient arrays and context parameters. Adds comprehensive PHPUnit tests for recipient management, context propagation, and shared inbox REST API behavior. Updates assertions to reflect new shared inbox semantics in REST controller tests.

pfefferle committed:
Remove post_author fallback in get_recipients

Eliminates backward compatibility logic that used post_author as a fallback recipient in Inbox::get_recipients. Corresponding test for this behavior has also been removed, reflecting the updated recipient determination logic.

matticbot committed:
Add changelog

pfefferle commented:
I will update the other handlers if that is how we want to proceed.

Jiwoon-Kim commented:
I’ve been thinking about it, and I believe it’s better to implement the outbox as a custom post type (CPT) and the shared inbox as a taxonomy.

Because the outbox already represents local activity types, if we implement remote activities as a custom post type (CPT) and set the inbox as a shared taxonomy between the two CPTs, it will allow us to receive both local and remote activities seamlessly.

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -27,32 +27,72 @@ class Inbox {
-	 * @param Activity|\WP_Error $activity The Activity object.
-	 * @param int                $user_id  The id of the local blog-user.
+	 * @param Activity|\WP_Error $activity   The Activity object.
+	 * @param int|array          $recipients The id(s) of the local blog-user(s).
```
Let's only document the new type and use `_deprecated_argument()` if `$recipients` is an int and convert it.

```php
if ( is_int( $recipients ) {
	\_deprecated_argument();

	$recipients = array( $recipients );
}
```

pfefferle replied:
Int is still fine.

obenland replied:
The code docs talk about backwards compatibility for ints, which indicates to me that they're no longer expected as parameters? The only place the function is called from, other than tests, is `Activitypub\Handler\Inbox::handle_inbox_requests()`, which also expects an array of user_ids.

If you want to keep it compatible with ints for potential future use, let's just remove the back compat language.

pfefferle replied:
AI thought I wanted to deprecate the int! Sorry :(

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -27,32 +27,72 @@ class Inbox {
+			function ( $value ) {
+				return null !== $value && false !== $value;
+			}
+		);
```
Is it possible to still have `null` and `false` values after `absint()`? We might be able to just remove this.

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -27,32 +27,72 @@ class Inbox {
+				return null !== $value && false !== $value;
+			}
+		);
+		\sort( $recipients );
```
Does this matter?

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -27,32 +27,72 @@ class Inbox {
+		if ( empty( $recipients ) ) {
+			return new \WP_Error(
+				'activitypub_inbox_no_recipients',
+				\__( 'No valid recipients provided', 'activitypub' ),
```
Is this a user-facing string? If not, we could avoid translation work by not sending it through `__()`.

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -222,4 +269,116 @@ public static function undo( $id ) {
+	 */
+	public static function has_recipient( $post_id, $user_id ) {
+		$recipients = self::get_recipients( $post_id );
+		return \in_array( (int) $user_id, $recipients, true );
```
```suggestion

		return \in_array( (int) $user_id, $recipients, true );
```

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -222,4 +269,116 @@ public static function undo( $id ) {
+		$result = \add_post_meta( $post_id, '_activitypub_user_id', $user_id, false );
+
+		if ( $result ) {
+			\clean_post_cache( $post_id );
```
`add_post_meta()` deletes its own meta cache, is there more that needs updating?

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -222,4 +269,116 @@ public static function undo( $id ) {
+		$result = \delete_post_meta( $post_id, '_activitypub_user_id', $user_id );
+
+		if ( $result ) {
+			\clean_post_cache( $post_id );
```
Same as above

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -222,4 +269,116 @@ public static function undo( $id ) {
+	 */
+	public static function remove_recipient( $post_id, $user_id ) {
+		$user_id = (int) $user_id;
+		// Allow 0 for blog user, but reject negative values.
```
```suggestion

		// Allow 0 for blog user, but reject negative values.
```

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -222,4 +269,116 @@ public static function undo( $id ) {
+		if ( ! self::has_recipient( $post->ID, $user_id ) ) {
+			return new \WP_Error(
+				'activitypub_inbox_not_recipient',
+				\__( 'User is not a recipient of this activity', 'activitypub' ),
```
Is this a user-facing string? If not, we could avoid translation work by not sending it through `__()`.

obenland commented on includes/handler/class-inbox.php:
```diff
@@ -21,20 +21,48 @@ class Inbox {
+	 * Supports both single user_id (int) and multiple user_ids (array).
+	 *
+	 * @param array              $data         The data array.
+	 * @param int|array          $user_ids     The id of the local blog-user, or array of user IDs.
```
Let's only document the new type and use `_deprecated_argument()` if `$user_ids` is an int and convert it.

pfefferle replied:
it needs to still support both, because the `activitypub_inbox` still only sends ints. no need to deprecate/change that.

obenland commented on includes/rest/class-inbox-controller.php:
```diff
@@ -168,28 +174,39 @@ public function create_item( $request ) {
-				 * @param int                $user_id  The user ID.
-				 * @param Activity|\WP_Error $activity The Activity object.
-				 */
-				\do_action( 'activitypub_inbox_' . $type, $data, $user_id, $activity );
```
Would adding the context not be enough for these actions to stay here? What's the benefit of moving them to run on `activitypub_inbox_shared`?

pfefferle replied:
To make the code more readable and to be able to unhook them... I would love to deprecate them over time, so I thought it might be nice to hide them a bit.

obenland replied:
I feel like it makes it much harder to read and understand. I'm happy for us to deprecate them, but I'm not sure they. need to be moved into a callback to do that?

pfefferle replied:
They should be out of the way... Implementers should (in the best case) not see them... The ideal solution (in my mind) would be to move them into a `deprecated.php` file... but I a have no strong feelings about that...

obenland commented on includes/class-migration.php:
```diff
@@ -1051,4 +1055,28 @@ public static function sync_jetpack_following_meta() {
 	}
+
+	/**
+	 * Cleanup inbox items for shared inbox migration.
```
```suggestion
	 * Clean up inbox items for shared inbox migration.
```
This is similar to setup and set up, where cleanup is the noun and clean up the verb.

obenland commented on includes/class-migration.php:
```diff
@@ -1051,4 +1055,28 @@ public static function sync_jetpack_following_meta() {
+	 * Deletes all existing inbox items to prepare for the new shared inbox structure
+	 * where activities are stored once with multiple recipients as metadata.
+	 */
+	private static function cleanup_inbox() {
```
```suggestion
	private static function clean_up_inbox() {
```

obenland reviewed (commented):
Mostly small stuff and some questions for me to better understand things

pfefferle committed:
Merge branch 'trunk' into improve-shared-inbox

pfefferle committed:
Refactor inbox cleanup and recipient handling

Renamed cleanup_inbox to clean_up_inbox for consistency and updated related references. Simplified recipient filtering logic in Inbox class and removed unnecessary cache cleaning after meta updates. Also removed translation wrappers from some error messages for improved clarity.

pfefferle committed:
Replace array_filter with array_values for recipients

Changed the sanitization of recipients to use array_values instead of array_filter, ensuring the array is re-indexed after removing duplicates and casting to integers.

pfefferle committed:
Update deprecation notes in inbox controller

Added 'unreleased' to deprecation comments for ActivityPub inbox actions to clarify support for activitypub_inbox_shared and avoid duplicate processing.

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -27,32 +27,65 @@ class Inbox {
+		}
+
+		// Sanitize recipients.
+		$recipients = \array_map( 'absint', $recipients );
```
```suggestion
		// Sanitize recipients.
		$recipients = \array_map( 'absint', (array) $recipients );
```

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -27,32 +27,65 @@ class Inbox {
+		$existing = self::get_by_guid( $activity->get_id() );
+
+		// If activity exists, add new recipients to it.
+		if ( ! \is_wp_error( $existing ) && $existing instanceof \WP_Post ) {
```
```suggestion
		if ( $existing instanceof \WP_Post ) {
```

obenland commented on includes/collection/class-inbox.php:
```diff
@@ -88,6 +121,13 @@ public static function add( $activity, $user_id ) {
 		}
 
+		// Add recipients as separate meta entries after post is created.
+		if ( ! \is_wp_error( $id ) && $id > 0 ) {
```
```suggestion
		if ( ! \is_wp_error( $id ) ) {
```
I'd expect post Ids to always be more than 0.

obenland commented on includes/handler/class-inbox.php:
```diff
@@ -21,20 +21,48 @@ class Inbox {
+	 * @param string             $context      The context of the request (Inbox_Collection::CONTEXT_INBOX or Inbox_Collection::CONTEXT_SHARED_INBOX).
+	 */
+	public static function handle_inbox_requests( $data, $user_ids, $type, $activity, $context = Inbox_Collection::CONTEXT_INBOX ) {
+		$inbox_id = null;
```
Could this be `$result`? It feels a bit less confusing than setting `$inbox_id` to WP_Errors and checking `$inbox_id` to see whether it should be added to the Inbox or not. Although it doesn't feel super great either.

obenland commented on includes/handler/class-inbox.php:
```diff
@@ -59,23 +86,39 @@ public static function handle_inbox_requests( $data, $user_id, $type, $activity
 		}
 
+		// Normalize user_id to array for action hooks.
+		$user_ids = is_array( $user_ids ) ? $user_ids : array( $user_ids );
```
```suggestion
		$user_ids = (array) $user_ids;
```
This might be good to do at the top of the function so `Inbox_Collection::add()` can take advantage of it, too.

obenland commented on includes/rest/class-inbox-controller.php:
```diff
@@ -156,9 +159,12 @@ public function create_item( $request ) {
 			foreach ( $recipients as $user_id ) {
-				// Check user-specific blocks for this recipient.
-				if ( Moderation::activity_is_blocked_for_user( $activity, $user_id ) ) {
+				if ( ! Moderation::activity_is_blocked_for_user( $activity, $user_id ) ) {
```
Regardless of what we decide about the fait of `trigger_inbox_actions()`, let's keep checking for something truthy here.

pfefferle committed:
Refactor inbox action triggering in REST controller

Moved the logic for triggering ActivityPub inbox actions directly into the recipient loop, removing the separate trigger_inbox_actions() method. This simplifies the code and reduces indirection for handling allowed recipients.

pfefferle committed:
Update includes/handler/class-inbox.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/collection/class-inbox.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/collection/class-inbox.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/collection/class-inbox.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Refactor inbox recipient filtering and hooks

Moved the 'activitypub_rest_inbox_disallowed' action to trigger immediately when a recipient is blocked, and streamlined the logic to improve clarity. This change ensures that hooks are fired in the correct context for both allowed and disallowed recipients.

pfefferle committed:
Rename inbox_id variable to result in handle_inbox_requests

Replaces the variable name $inbox_id with $result throughout the handle_inbox_requests method for clarity, as it can hold either a WP_Error or an inbox item ID. Updates all references and related documentation accordingly.

pfefferle committed:
Refactor inbox action hook parameter names and usage

Renamed the fifth parameter of the 'activitypub_handled_inbox_*' action from $inbox_id to $result for clarity. Updated all related usages and tests to reflect this change, including increasing the number of accepted parameters from 5 to 6 in test action hooks. Removed an unused action registration in Inbox_Controller.

pfefferle committed:
Merge branch 'trunk' into improve-shared-inbox
